### PR TITLE
Move user data export to background tasks

### PR DESCRIPTION
**Pull request by amCap1712**
User data exports take a long time and get disrupted because the request times out during the process. To fix the issues, do the exports to an archive in the background and allow the users to download the archive later. Archives are auto-deleted after a fixed period.

TODO:
1. Create docker volume to store exports and update consul config in production.
2. Discuss with @zas where to store the export and possibility of using gateways for a speed up.
3. ~~Discuss with @MonkeyDo to add a screen for users to view their completed export and download it, also update the existing export frontend.~~
4. Add tests.

image

image

### CODE REVIEW

File: admin/sql/updates/2024-08-06-add-background-export.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
+ALTER TYPE background_tasks_type ADD VALUE 'export_user';
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply admin/sql/updates/2024-08-06-add-background-export.sql:1]: Sure, will change. FWIW, my intention was to keep it all in one export at least for the forseeable future.

[the remark this answers is not in the document]
Comment (MonkeyDo) [reply admin/sql/updates/2024-08-06-add-background-export.sql:1]: I don't think we need anything else for now, just wanted to make sure we word it right for potential future features

**Comment by MonkeyDo** [comment]
Initial suggestions for the UI:

I reckon we want to show any ongoing export (along with progress, download links eventually) on the same page it was initially requested, at the top of the page.

Maybe something like this?
image

image

@Aerozol might have some valuable input and suggestions?

**Comment by Aerozol** [comment]
I don't see any issue with that UI. Additional thoughts:
- If users can leave and come back later we could include the date + time of when the export was requested (or the date/time of the last listen in the export could be more useful)
- If requesting an export is about to overwrite a current one we might want to indicate that (or just a small note like "will reset any current exports").

File: admin/sql/updates/2024-08-06-add-background-export.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (amCap1712) [reply admin/sql/updates/2024-08-06-add-background-export.sql]: Fixed.

**Comment by amCap1712** [comment]
@MonkeyDo I have endpoints to retrieve existing and in progress exports, and also the current progress of a given export. I think we should also add a table of available exports to the page.

**Comment by MonkeyDo** [comment]
@amCap1712 I believe the front-end parts are pretty much done (added some screenshots to the PR description), although I will need to do some more manual testing as I think currently the zip files are not being saved in my local setup, and trying to download said file results in a server error.

I suppose this is point 1 in the to-do list, setting up a volume to export the files at.

I also leave the ball in your court to look at the failing server tests, which I reckon is point 4 in the to-do list.

If your head in under the water with too many projects to finish, let me know and we can try to rope Ansh in to help hammer the last nails in.

**Comment by amCap1712** [comment]
@MonkeyDo I tested the PR on beta and it works good. Can you please update this PR to allow deletion of exports and let users create a new export and if an exiting and complete export exists.

**Comment by MonkeyDo** [comment]
@amCap1712 I hadn't seen all the functionality XD
All caught up now, you can delete/cancel exports and request new ones when there are none in progress (i.e even with a completed or failed export you can request a new one)
Updated the screenshots in the PR description.

**Comment by MonkeyDo** [comment]
I will add that currently when I try to delete an export, I get this error in the console/flask app:
```
web-1             |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 870, in full_dispatch_request
web-1             |     return self.finalize_request(rv)
web-1             |            ^^^^^^^^^^^^^^^^^^^^^^^^^
web-1             |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 889, in finalize_request
web-1             |     response = self.make_response(rv)
web-1             |                ^^^^^^^^^^^^^^^^^^^^^^
web-1             |   File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 1197, in make_response
web-1             |     raise TypeError(
web-1             | TypeError: The view function did not return a valid response. The return type must be a string, dict, list, tuple with headers or status, Response instance, or WSGI callable, but it was a APINotFound.
```

Let me know if I can help debug

**Comment by amCap1712** [comment]
@MonkeyDo I think I have fixed the error. I also made some changes to the frontend to avoid page reloads on deletion and cancellation of export. Also, some minor UI/UX fixes. The latest PR is live on beta, please test and feel free to merge if everything seems fine.

**Comment by Aerozol** [comment]
Nice work with the UI with the info drop down

**Comment by MonkeyDo** [comment]
> @MonkeyDo I think I have fixed the error. I also made some changes to the frontend to avoid page reloads on deletion and cancellation of export. Also, some minor UI/UX fixes. The latest PR is live on beta, please test and feel free to merge if everything seems fine.

Great, thank you! I have one more minor thing to add, a small button to refresh an ongoing export to get the latest status.
I'll merge after that.

**Comment by MonkeyDo** [comment]
@amCap1712 Looks great now ! Done on my side.
Just the one failing server test left, not sure why the file is still in the data export directory, but I reckon the code in `delete_export_archive` does nothing with the returned `filename`, so probably doesn't actually deleted the export zip file maybe?

**Comment by amCap1712** [comment]
fixed now, thanks!

**Pull request merged**